### PR TITLE
Fix MPS scatter rank mismatch for scalar-index updates

### LINKED ISSUE #39: `mps.scatter` crashes with rank mismatch (2 != 3) when update tensor has implicitly squeezed dimensions
I am encountering a crash when using `jax.lax.scatter` (via the `.at[].set()` syntax). The error occurs when the `update` tensor has a lower rank than the `destination` tensor, specifically when a dimension is implicitly squeezed (e.g., updating a `(1, 256)` slice into a `(1024, 1, 256)` tensor).

The same pattern never gave me any problems on CPUs or TPUs, so I guess it should be supported.

**Error Message**
```text
(mpsFileLoc): .../MPSGraphUtilities.mm:190:0: error: 'mps.scatter' op invalid input tensor shapes, updates rank and result rank must be equal, 2 != 3
(mpsFileLoc): .../MPSGraphUtilities.mm:190:0: note: see current operation: %6 = "mps.scatter"(%arg0, %arg2, %4, %5) <{mode = #mps.scatter_mode<set>}> : (tensor<10x1x4xbf16>, tensor<1x4xbf16>, tensor<1xsi32>, tensor<si32>) -> tensor<10x1x4xbf16>
/AppleInternal/.../MPSGraphExecutable.mm:1347: failed assertion `original module failed verification'
```

**Minimal Reproduction (thank you codex)**
```python
import jax
import jax.numpy as jnp

def test_mps_scatter_rank_mismatch():
    # Ensure we are on MPS
    print(f"JAX Backend: {jax.default_backend()}")
    
    # Destination Tensor: Rank 3
    # Shape: (Time, Heads, Head_Dim)
    dest = jnp.zeros((10, 1, 4), dtype=jnp.bfloat16)
    
    # Update Tensor: Rank 2
    # Shape: (Heads, Head_Dim)
    # logic: We are updating a single timestep index, so the time dim is removed.
    update = jnp.ones((1, 4), dtype=jnp.bfloat16)
    
    # Index to update
    idx = 0
    
    print(f"Dest shape: {dest.shape} (Rank {len(dest.shape)})")
    print(f"Update shape: {update.shape} (Rank {len(update.shape)})")
    print("Attempting .at[].set operation...")
    
    # CRASHES HERE
    out = dest.at[idx].set(update)
    
    print("Success!")
    print(out.shape)

if __name__ == "__main__":
    test_mps_scatter_rank_mismatch()
```
I verified that the code above works fine on my CPU. 

**Version**
jax-mps: 0.9.5

I'm happy to attempt to fix this myself if you want @tillahoffmann. Would be a good opportunity for me to learn some jax internals, but if you want it fixed quickly, it might be better not to wait for me to understand all the details.

### PR DESCRIPTION
**Pull request by dominik3141**
Fixes #39.

This crash was caused by a mismatch between StableHLO scatter semantics and what MPSGraph requires.

For scalar-index patterns like `x.at[idx].set(update)`, StableHLO may represent `update` with one fewer dimension (the scattered axis is implied via `inserted_window_dims`). `MPSGraph scatterWithDataTensor` is stricter and expects update rank to match operand/result rank, which led to the reported `2 != 3` failure.

This PR updates the supported single-axis scatter lowering in `src/pjrt_plugin/ops/shape_ops.mm`:
- when `updates.rank + 1 == input.rank`, it reshapes updates to reinsert a singleton axis at `scatterAxis` before calling MPS scatter.
- behavior for already rank-aligned updates is unchanged.

It also adds a regression case in `tests/configs/slice.py` (`scalar_index_set_rank_squeezed_update`) matching the issue pattern: destination `(10, 1, 4)`, scalar index, update `(1, 4)` with `bfloat16`.

I verified the issue repro runs on MPS after this change, and the slice-focused op tests still pass.

### CODE REVIEW

**Comment by dominik3141** [comment]
Now handles rank mismatches > 1 for scatter.

I also wrote a handler for a scatterNd pattern (a badly written test (by myself) pointed me towards that pattern not existing).  And added the necessary gather counterpart handler so that autodiff works.

**Comment by dominik3141** [comment]
I think this should be ready to be merged @tillahoffmann, just make sure to double check my code

File: src/pjrt_plugin/ops/shape_ops.mm [hunk 1]
```diff
@@ -465,6 +466,56 @@ static ProcessResult HandleGather(HandlerContext& ctx) {
...
+    // - indices is a 1D vector of length input_rank
+    // - index_vector_dim is 0, all indices are included in the vector
+    // - start index map covers all operand dimensions
+    // - no offset dimensions and no collapsed slice dimensions (scalar result)
+    if (indexVectorDim == 0 && indicesRank == 1 &&
+        startIndexMap.size() == operand.shape.count &&
+        offsetDims.empty() &&
+        [indicesShape[0] integerValue] == (NSInteger)operand.shape.count) {
+        bool hasFullCollapsedSliceDims = collapsedSliceDims.size() == operand.shape.count;
+        if (hasFullCollapsedSliceDims) {
```
[the remark this answers is not in the document]
Comment (dominik3141) [reply src/pjrt_plugin/ops/shape_ops.mm]: Fair point
